### PR TITLE
Klocwork - solving the nullptr issue

### PR DESCRIPTION
**Pull request by CoralMalachi**
Resolving a klocwork issue which leads to many errors.

Since a null pointer may cause undefined behavior when the program attempts to
access memory through it, each time we tried to access to an object that
may be null without validates it is not, the klocwork created an issue
for it with the following message:

`Pointer *** returned from call to function *** may be NULL and will be dereferenced at ...`

To solve all of those issues, a validation check is added each time we try
to call a member function on an object that can be an invalid object (a
null pointer doesn't point to a valid object).

Signed-off-by: CoralMalachi <coral.malachi@intel.com>

### CODE REVIEW

**Comment by CoralMalachi** [comment]
Please do not merge yet.

File: agent/src/beerocks/monitor/monitor_stats.cpp
Comment (arnout) [inline agent/src/beerocks/monitor/monitor_stats.cpp:189]: The rest of the commit is an almost mechanical change `foo == nullptr` -> `!foo`. Therefore it would be better to do this bit in a separate commit.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (arnout) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:1881]: Same here, better as a separate commit.

File: agent/src/beerocks/slave/platform_manager/platform_manager_thread.cpp
Comment (arnout) [inline agent/src/beerocks/slave/platform_manager/platform_manager_thread.cpp:1315]: You changed the semantics here:

```suggestion
    if ((!sd) && (!(sd = get_backhaul_socket()))) {
```

File: agent/src/beerocks/slave/son_slave_thread.cpp
Comment (arnout) [inline agent/src/beerocks/slave/son_slave_thread.cpp:1569]: Again, better in a separate commit.

File: common/beerocks/tlvf/src/include/beerocks/tlvf/beerocks_header.h
Comment (arnout) [inline common/beerocks/tlvf/src/include/beerocks/tlvf/beerocks_header.h:48]: This is exactly the same as the original... Does klocwork really barf on `?:` expressions? Or isn't it be able to recognise that `actionhdr()` always returns the same value? Hm, probably the latter... Would it help if we would add `__attribute__((pure))` to the actionhdr declaration?

Oh, and we probably need to make all those functions const as well. But that's unrelated.

File: controller/src/beerocks/master/db/network_map.cpp
Comment (arnout) [inline controller/src/beerocks/master/db/network_map.cpp:137]: I realize that you didn't change this code, but something is fishy here - the first assignment is redundant...

File: controller/src/beerocks/master/db/network_map.cpp
Comment (arnout) [inline controller/src/beerocks/master/db/network_map.cpp:151]: Does klocwork really require this? To me, it would be more natural to do:

```
    data_start = (uint8_t *)response->buffer(0);
    if (!data_start) {
```

File: controller/src/beerocks/master/db/network_map.cpp
Comment (arnout) [inline controller/src/beerocks/master/db/network_map.cpp:352]: Why?

File: controller/src/beerocks/master/db/network_map.cpp
Comment (arnout) [inline controller/src/beerocks/master/db/network_map.cpp:434]: Hm, this pattern is repeating a couple of times. And actually it is a bit silly: if create_vs_message doesn't return nullptr, then we're guaranteed to have a beerocks_hrd and an actionhdr...

Maybe we should just extend create_vs_message with an optional std::shared_ptr<beerocks_message::cACTION_HEADER> argument that gets updated with the action header for the cases that need it?

File: controller/src/beerocks/master/db/network_map.cpp
Comment (arnout) [inline controller/src/beerocks/master/db/network_map.cpp:556]: Meh. This stuff was ugly already, and it doesn't get better after the change... Why aren't bml events generated like the rest of the code? @morantr @vitalybu 

At least, if it is not generated, it would be better to add a helper template function that does all of the parsing and checking. And get rid of the GET_MESSAGE_POINTER macro.

**Review by arnout — CHANGES_REQUESTED** [review]
Other than the bug in `main_thread::handle_arp_monitor` it looks good to me. However, because there's so much squashed into a single commit, I can't be sure that I didn't miss anything. Please split the commits.

File: controller/src/beerocks/master/db/network_map.cpp
Comment (CoralMalachi) [reply controller/src/beerocks/master/db/network_map.cpp:137]: Yes, I saw that too, @morantr can I remove line 136?

Comment (tomereli) [reply controller/src/beerocks/master/db/network_map.cpp:137]: For something like this you don't need approval from Moran.. the first assignment is simply redundant so just remove it

File: common/beerocks/tlvf/src/include/beerocks/tlvf/beerocks_header.h
Comment (CoralMalachi) [reply common/beerocks/tlvf/src/include/beerocks/tlvf/beerocks_header.h:48]: Yes, klocwork complainants whenever we try to access the return value of any method that returns a shared pointer unless we validate that it is not null

File: agent/src/beerocks/slave/platform_manager/platform_manager_thread.cpp
Comment (CoralMalachi) [reply agent/src/beerocks/slave/platform_manager/platform_manager_thread.cpp:1315]: nice catch

File: common/beerocks/tlvf/src/include/beerocks/tlvf/beerocks_header.h
Comment (tomereli) [reply common/beerocks/tlvf/src/include/beerocks/tlvf/beerocks_header.h:48]: But ` = actionhdr() ? ACTION_NONE : actionhd->action();` is checking exactly the same thing. Weird.

**Comment by arnout** [comment]
One more thing: could you add a commit that updates CODINGSTYLE and explains not to use `== nullptr`?

**Comment by CoralMalachi** [comment]
> One more thing: could you add a commit that updates CODINGSTYLE and explains not to use `== nullptr`?

Sure.

File: common/beerocks/tlvf/src/include/beerocks/tlvf/beerocks_header.h
Comment (arnout) [reply common/beerocks/tlvf/src/include/beerocks/tlvf/beerocks_header.h:48]: No, because there is no guarantee that actionhdr() will return the same value in both cases. That's why I suggested to declare it as pure - if klocwork knows the pure attribute, it can use it to determine that in fact the same value will be returned in both cases.

File: framework/tlvf/src/src/CmduMessage.cpp
Comment (arnout) [inline framework/tlvf/src/src/CmduMessage.cpp:21]: This is a bit annoying. If getCmduHeader() doesn't return nullptr, then prevClass() is guaranteed to return non-null because the cmdu header is the first class in the list. So this additional check is entirely redundant, and it's just a matter of insufficient intelligence in the tool (i.e., a false positive).

In this case, we don't actually make use of the getCmduHeader result, so we can get away with just removing the first check.

File: controller/src/beerocks/master/db/network_map.cpp
Comment (CoralMalachi) [reply controller/src/beerocks/master/db/network_map.cpp:434]: I'll do it in next commit, great idea

File: controller/src/beerocks/master/db/network_map.cpp
Comment (morantr) [reply controller/src/beerocks/master/db/network_map.cpp:137]: Actually it is a serious bug now that I look at it. `num_of_nodes` is defined as reference of `response->node_num()` at the function beginning. On that block, we create a new message but we still use the reference to the previous message. It is accidentally working since the `num_of_nodes` location on the buffer is not changed when creating a new message.

I think that regardless of deleting the redundant line, we need also to use `response->node_num()` directly instead of the reference, and delete the reference declaration at the function beginning.

File: controller/src/beerocks/master/db/network_map.cpp
Comment (morantr) [inline controller/src/beerocks/master/db/network_map.cpp:180]: I don't think it was right to replace it also on debug prints.

**Review by morantr — CHANGES_REQUESTED** [review]

**Pull request closed without merging**
